src/Grpc.NetCore.HttpClient/Internal/HttpContentClientStreamWriter.cs

```diff
@@ -0,0 +1,127 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+
+namespace Grpc.NetCore.HttpClient.Internal
+{
+    internal class HttpContentClientStreamWriter<TRequest, TResponse> : IClientStreamWriter<TRequest>
+    {
+        private readonly GrpcCall<TRequest, TResponse> _call;
+        private readonly Task<Stream> _writeStreamTask;
+        private readonly TaskCompletionSource<bool> _completeTcs;
+        private readonly object _writeLock;
```

Comment:
SemaphoreSlim?

Reply:
The lock is used when a write starts, it doesn't lock the entire write. It isn't used for async locking.